update kube-webhook-certgen to v1.5.1

arikgrahl opened the pull request:
<!--- Provide a general summary of your changes in the Title above --->
<!--- Please don't @-mention people in PR or commit messages (do so in an additional comment). --->

## What this PR does / why we need it:

- update [kube-webhook-certgen](https://github.com/jet/kube-webhook-certgen) to [v1.5.1](https://github.com/jet/kube-webhook-certgen/releases/tag/v1.5.1)
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] CVE Report (Scanner found CVE and adding report)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation only

## Which issue/s this PR fixes
<!--
(optional, in `fixes #<issue number>` format, will close that issue when PR gets merged):

fixes #
-->
–

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/kubernetes/ingress-nginx/blob/main/CONTRIBUTING.md) guide
- [ ] I have added unit and/or e2e tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Added Release Notes.

## Does my pull request need a release note?

```release-note
Update [kube-webhook-certgen](https://github.com/jet/kube-webhook-certgen) to [v1.5.1](https://github.com/jet/kube-webhook-certgen/releases/tag/v1.5.1)
```

rikatz commented:
We have our own fork of kube web certgen. Please check images/ folder

arikgrahl commented:
Thanks for pointing out the fork.
The README.md in the images folder says

> Only the nginx image is meant to be published
> Other images are used as examples for a feature of the ingress controller or to run e2e tests

In the latest version of the ingress-nginx helm chart, the image `registry.k8s.io/ingress-nginx/kube-webhook-certgen:v20220916-gd32f8c343@sha256:39c5b2e3310dc4264d638ad28d9d1d96c4cbb2b2dcfb52368fe4e3c63f61e10f` is used.
Does this image result from the said fork?

The endpoint https://registry.k8s.io/v2/ingress-nginx/kube-webhook-certgen/tags/list lists the newer tag `v20221220-controller-v1.5.1-58-g787ea74b6` among other, to which this PR updates the helm chart.
Regardless of this, for better transparency, an image tag with an actual application version number would be favored, which the image tag of this update would satisfy.

strongjz commented:
@arikgrahl Please rebase your PR and sign the CLA. 

thank you for your work on this. Tracking and keeping all of our images up to date is not an easy task so we appreciate this work. 

-James

arikgrahl commented:
I have rebased my PR. In that course I also fixed the email address in the commit so that the CLA check for the previously signed CLA passed as well. Thank you for pointing this out!

tao12345666333 reviewed (commented):
/ok-to-test
/hold

There may be a reason for this image not being updated.  I will think about the reason

arikgrahl commented:
@tao12345666333, any updates on this?

To me it is unclear which exact upstream version is behind the tag `v20220916-gd32f8c343`.
This makes it personally hard to predict which changes are introduced by this update to the tag `v20221220-controller-v1.5.1-58-g787ea74b6`.
At least the updated tag has some potential link to an appropriate upstream version ([v1.5.1](https://github.com/jet/kube-webhook-certgen/releases/tag/v1.5.1)), which could clear up any uncertainties in this regard for future updates.

This being said, the decision of accepting this PR will not be easier in the future without further background information about the image tags mentioned above.
Possibly the test suite is sufficient to identify any breaking changes for this update?
Otherwise, I will be happy to volunteer for specific testing on this update.

Thank you in advance!

strongjz commented:
I'm working on a 1.6.5 release this week, and I just upgraded all the images to golang 1.20 and alpine 3.17.2

https://github.com/kubernetes/ingress-nginx/pull/9690/files#diff-a4e1312dfd8eddfe3111f1c850a112c92040b42c8214d434cc1bdd44aa4c23e3

https://github.com/kubernetes/k8s.io/pull/4929

So there be an even newer version of the cert image. 

We can close this issue in anticipation of the new release. 

Thank you,
James

/close

arikgrahl commented:
@strongjz, thank you for this upcoming release, i am looking forward to it!

However, the upcoming release will [update components in the images folder](https://github.com/kubernetes/ingress-nginx/pull/9690/files#diff-a4e1312dfd8eddfe3111f1c850a112c92040b42c8214d434cc1bdd44aa4c23e3), among others.
[As already pointed out](https://github.com/kubernetes/ingress-nginx/pull/9557#issuecomment-1409223924), the README.md in this folder states:

> Only the nginx image is meant to be published
> Other images are used as examples for a feature of the ingress controller or to run e2e tests

If I am not missing something, this means that the helm chart, for which this PR is aiming, will remain with the old version.
I guess this was already initially confused by @rikatz.

Please reconsider including this minor update if I am not wrong.

Thank you again for your work!

longwuyuan commented:
@arikgrahl when a new kube-webhook-certen image is built by cloudbuild, it will be promoted to be published from the prod registry and the manifests that refer to the kube-webhook-certgen image will get a update on the tag and the digest.

arikgrahl commented:
@longwuyuan, thank you for clarifying.
I thought that the process would be like this, but noticed a few other tags, which seemed newer than the one in the helm chart manifests.

```
curl -s -L https://registry.k8s.io/v2/ingress-nginx/kube-webhook-certgen/tags/list | jq .tags
```
```json
[
  "sha256-2188def723c8fcb2635fbe54607ac3b4efba581c4fcc3d2279ec9c57da1b2f6f.sig",
  "sha256-39c5b2e3310dc4264d638ad28d9d1d96c4cbb2b2dcfb52368fe4e3c63f61e10f.sig",
  "sha256-4d99688e557396f5baa150e019ff7d5b7334f9b9f9a8dab64038c5c2a006f6b5.sig",
  "sha256-549e71a6ca248c5abd51cdb73dbc3083df62cf92ed5e6147c780e30f7e007a47.sig",
  "sha256-ccdb2382b53ed42fd76bfbff90a5b8a5815af0b3b8c6dfb9118a34770d9cac11.sig",
  "v1.0",
  "v1.1",
  "v1.1.1",
  "v1.2.0",
  "v1.3.0",
  "v20220916-gd32f8c343",
  "v20221219-controller-v1.5.1-49-ge3e0d9c1f",
  "v20221220-controller-v1.5.1-58-g787ea74b6"
]
```

Still it remains unclear, where the actual images come from and which exact (upstream) version is behind the respective image tags.
If the components from the [image folder](https://github.com/kubernetes/ingress-nginx/tree/main/images) are utilized in the helm chart manifests, the [respective notice](https://github.com/kubernetes/ingress-nginx/blob/b87645fe5318dc6fe9b00b3d9c6c25b5098d9e0e/images/README.md) should reflect this.
In any case, I am happy to enhance the documentation to create as much transparency as possible in this great open source project.

longwuyuan commented:
The explaining is long and complicated. For practical ends, be assured that regardless of tags you see, a documented & recommended will always get you the kube-webhook-certgen code bits that you need.

There used to be a upstream. Now there is no upstream. We painfully forked and also carefully maintain the fork. Docs and users are user-facing but kube-webook-certgen is the woodworks, hence not much is needed except for it to work as expected.

Latest is as James said. He will trigger build of a new kube-webhook-certgen and code will be updated to the new tag and digest.

arikgrahl commented:
@longwuyuan, thank you so much for shedding light on this and patiently answering my questions!

Thank you for the great work in this project, I appreciate it very much.

longwuyuan commented:
> @longwuyuan, thank you so much for shedding light on this and patiently answering my questions!
> 
> Thank you for the great work in this project, I appreciate it very much.

:pray:  community
